### PR TITLE
[TS migration] Migrate 'EmojiPickerAction.js' lib to TypeScript

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

Migration of `EmojiPickerAction.js` to Typescript.

### Fixed Issues
$ https://github.com/Expensify/App/issues/24798
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console
- [x] Verify that `EmojiPicker` is working correctly

### Offline tests
N/A

### QA Steps
N/A

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/46095609/ec4b4093-4bc1-4290-a736-1579d62e083f

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/46095609/ac5593bb-ba8a-4853-97bb-25a0cc719f41

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/46095609/0246b79c-8cbb-49ba-892a-9543c8e26c1d

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/46095609/f83ce121-8154-4ef0-92ee-11ff44925e93

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/46095609/e604546c-8709-425f-a33e-2460e3908253

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/46095609/e4a8b868-d5cb-487e-87cd-17f7883baf34

</details>

### CODE REVIEW

**Review by ntdiary — APPROVED** [review]

File: src/libs/actions/EmojiPickerAction.ts
Comment (hayata-suenaga) [inline src/libs/actions/EmojiPickerAction.ts:19]: I just want to mention I have no idea why we keep references to components outside the component tree.

I asked [a question](https://expensify.slack.com/archives/C01GTK53T8Q/p1694031987704399) about this in the #expensify-open-source slack channel.

This is not part of this PR, but we probably should stop doing this pattern.

File: src/libs/actions/EmojiPickerAction.ts
Comment (hayata-suenaga) [inline src/libs/actions/EmojiPickerAction.ts:11]: Why are we defining the component type in this file?

File: src/libs/actions/EmojiPickerAction.ts
Comment (hayata-suenaga) [inline src/libs/actions/EmojiPickerAction.ts:12]: What is the purpose of `anchorOrigin` if it always takes exactly same object?

File: src/libs/actions/EmojiPickerAction.ts
Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:19]: Yes, mee too. It's really odd to me 👍

File: src/libs/actions/EmojiPickerAction.ts
Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:12]: `AnchorOrigin` is of type:
```ts
type AnchorOrigin = {
    horizontal: 'left' | 'center' | 'right';
    vertical:  'top' | 'center' | 'bottom';
};
```

Soo, It isn't always the same object 😄

File: src/libs/actions/EmojiPickerAction.ts
Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:11]: It's not the component type, it's what is defined in `useImperativeHandle` [here](https://github.com/software-mansion-labs/expensify-app-fork/blob/d3b1cf5962e638007755ec2f60fc0715345a5602/src/components/EmojiPicker/EmojiPicker.js#L119). Do you think the name is misleading? @hayata-suenaga

Comment (hayata-suenaga) [reply src/libs/actions/EmojiPickerAction.ts:11]: I mean this is component API right? the methods exposed the component through `useImperativeHandle`?

I think we should define this inside the component file instead of defining this here where we store the reference to the  component.

I know we haven't started migration of component files, but I think we should start thinking about where type definitions should belong and how to coordinates the migration when the type needed are not defined yet.

File: src/libs/actions/EmojiPickerAction.ts
Comment (hayata-suenaga) [reply src/libs/actions/EmojiPickerAction.ts:12]: ahhhh `CONST.MODAL.ANCHOR_ORIGIN_HORIZONTAL` is actually an object! sorry I missed that 🙇

File: src/libs/actions/EmojiPickerAction.ts
Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:11]: I agree that `EmojiPicker` type should be defined next to `EmojiPicker` component. But until it's not migrated we have couple options:
- Put this PR on hold until we get to `EmojiPicker` component
- Merge this now, and move the type when `EmojiPicker` component is migrated (we can add a comment on the issue to remember about this)
- Define types.ts file with `EmojiPicker` type next to the component

Generally I think we won't be able to avoid such situations and sometimes we have to define types in undesirable places. At some point of the migration we will have to clean a bit, for example we'll have to clean style types or define shared types that are often used - until then we have to accept that we will repeat or litter the code sometimes.

cc @fabioh8010 @hayata-suenaga

Comment (fabioh8010) [reply src/libs/actions/EmojiPickerAction.ts:11]: I'm confortable with option two, **and adding a TODO comment above the type as well**:

```
Merge this now, and move the type when EmojiPicker component is migrated (we can add a comment on the issue to remember about this)
```

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [inline src/libs/actions/EmojiPickerAction.ts:31]:
```suggestion
function showEmojiPicker(onModalHide = () => {}, onEmojiSelected = () => {}, emojiPopoverAnchor = undefined, anchorOrigin = undefined, onWillShow = () => {}, id = undefined) {
```

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [inline src/libs/actions/EmojiPickerAction.ts:42]:
```suggestion
function hideEmojiPicker(isNavigating: boolean) {
```

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [inline src/libs/actions/EmojiPickerAction.ts:69]:
```suggestion
function resetEmojiPopoverAnchor() {
```

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [reply src/libs/actions/EmojiPickerAction.ts:11]: Maybe we will have to rename this type to e.g. `EmojiPickerRef` as well, to avoid conflicts.

Comment (hayata-suenaga) [reply src/libs/actions/EmojiPickerAction.ts:11]: > Merge this now, and move the type when EmojiPicker component is migrated (we can add a comment on the issue to remember about this)

I also like the option #2.

> Generally I think we won't be able to avoid such situations and sometimes we have to define types in undesirable places. At some point of the migration we will have to clean a bit, for example we'll have to clean style types or define shared types that are often used

I agree. It's just we must remember all the cleanups we have to do. I don't have the capacity to manage this, so I'd appreciate if SWM and CallStack and develop a strategy to coordinate and keep track the cleanup efforts we have to do. 🙇 (like what you suggested in option 2 i.e. commenting future migration issues to remember cleanup items)

File: src/libs/actions/EmojiPickerAction.ts
Comment (hayata-suenaga) [reply src/libs/actions/EmojiPickerAction.ts:31]: @blazejkustra @fabioh8010 did we have a style guide on whether to add `void` to a function returning undefined?

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [reply src/libs/actions/EmojiPickerAction.ts:11]: We can create some cleanup and review issues for some of the categories of files (I kinda have been done this cleanup for `styles` already) , e.g. `styles`, `models`, `libs`, `hooks`, `hoc's`, `components` and `pages` I think are good candidates for these clean ups. After each migration of all issues of that category we will create a new issue to review all past work and ensure consistency. I can add these "checkpoints" on our spreadsheet, WDYT? @hayata-suenaga @blazejkustra

Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:11]: @fabioh8010 That sounds good to me 😄 Add it to the spreadsheet please, and we will get back to this.

> I don't have the capacity to manage this, so I'd appreciate if SWM and CallStack and develop a strategy to coordinate and keep track the cleanup efforts we have to do. 🙇

Sure thing @hayata-suenaga, that's on me and @fabioh8010, we will coordinate such cleanups!

Comment (hayata-suenaga) [reply src/libs/actions/EmojiPickerAction.ts:11]: Let me know if issues need to be created in App 🙇

File: src/libs/actions/EmojiPickerAction.ts
Comment (blazejkustra) [reply src/libs/actions/EmojiPickerAction.ts:31]: We have a section that says we should rely on inferred type in such occasions, I'll change it.

**Comment by blazejkustra** [comment]
@fabioh8010 @hayata-suenaga Ready for your review!

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [reply src/libs/actions/EmojiPickerAction.ts]: @hayata-suenaga @blazejkustra I created new rows in the spreadsheet about these review issues, here are the lines:

- 57
- 261
- 277
- 295
- 486
- 550

@hayata-suenaga Could you please create the issues? Thank you!

**Review by fabioh8010 — APPROVED** [review]
LGTM, CK still need to do the checklist.

**Comment by hayata-suenaga** [comment]
I'll wait for the checklist to be completed before I do review 🙇

**Comment by fabioh8010** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/20051562/7f8327f4-ef26-4149-ad5f-e124da7bb6bb


https://github.com/Expensify/App/assets/20051562/a5fa29e9-362f-4330-acee-2622208e5474



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/20051562/e01e4d85-fb5b-4fc2-826f-e90bb19d2e8e



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/20051562/de0b70c7-7c28-4022-93e9-dd5a2b2364e4



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/20051562/95749bb4-20fd-430c-925b-a9ff833729d9



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/20051562/3f47cbd6-4b3c-4488-8a67-665d2c08b0b3



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/20051562/98ecca2f-24c8-46e6-aa7a-085b53a8465a



</details>

**Review by fabioh8010 — APPROVED** [review]

File: src/libs/actions/EmojiPickerAction.ts
Comment (fabioh8010) [reply src/libs/actions/EmojiPickerAction.ts]: @hayata-suenaga bump

**Review by hayata-suenaga — APPROVED** [review]

**Pull request merged**
